### ISSUE TITLE
Intl.Locale API

### ISSUE THREAD

**Issue by zbraniecki** [tc39/ecma402#46]
I started working on a spec and polyfill for Intl.Locale API.

I should have some proposal for this next week. Currently I have five functions on it:
 * isStructurallyValidLanguageTag(locale)
 * canonicalizeLocaleList(locales)
 * resolveLocale(availableLocales, requestedLocales, options, relevantExtensionKeys, localeData)
 * prioritizeAvailableLocales(availableLocales, requestedLocales, defaultLocale)
 * getDirection(locale)

The rationale for those five are:

* isStructurallyValidTag
This allows testing language tag for sanitation/validation purposes.

* canonicalizeLocaleList
This is the core function allowing all language negotiation operations

* resolveLocale

this exposes the core function used by all Intl formatters that allows people to write their own custom formatters or polyfills.

* prioritizeAvailableLocales

This is basic language negotiation function for all localization libraries. It takes requestedLocales list (navigator.languages or custom), availableLocales (app provided locales), and a fallback defaultLocale (default app locale, not default host env locale) and returns the list of prioritized available locales that best matches user requested locales.

* getDirection

Simple function allowing l10n libraries to set html.dir based on their negotiated locale.

@caridy  - One question I'd like to get help with is about using `Set` instead of `Array`. In all cases where we carry a list of locales, we always want it deduplicated and `Set` does just that. It would be sweet to use it instead of an `Array`, but it's API seems to be significantly less user friendly than that of the `Array` - it's impossible to just retrieve `locales[0]`, it's not easy to `map` etc.
Should I stick to `Array` or try to use `Set`?

**Comment by caridy** [comment]
Two things:

 * You can certainly use a Set internally as part of the logic, but let's stick to an array for arguments, I haven't seen any precedent of using a Set as arguments in Ecma.
 * Instead of implementing a new polyfill (once we settle on the api) you can just submit a PR for Intl.js, which already implements those abtract operations.

**Comment by zbraniecki** [comment]
sure! thanks for the feedback :)

**Comment by zbraniecki** [comment]
@caridy - I started playing with Intl.js and ended up separating `src/locale.js` out of `src/core.js` and then carved out the common functions into `src/utils.js` because `src/core.js` is huuge :)
Not sure if you're interested in any sort of refactor like that - if not, I'll bring it back into `src/core.js`.

Branch is here: https://github.com/zbraniecki/Intl.js/tree/locale

I also wrote a simple polyfill just for `Intl.Locale` based on Gecko implementation - https://github.com/zbraniecki/IntlLocale

Let's now talk about the API and get it agreed upon!

**Comment by caridy** [comment]
@zbraniecki I added the module system into Intl.js to be able to split it into small chunks, I didn't get to split core into small chunks. In other words, go for it, we will appreciate it. Just open a PR there, and we can discuss the details. 

In the other hand, I'm more interested on the API discussion :)

**Comment by zbraniecki** [comment]
Cool. Do you have any comments on the API I proposed in comment 1? I'd like to make Locale a good unified language negotiation API for other Intl APIs and localization frameworks.

I believe that with those five functions are enough to achieve that. On top of that, all of that code except of `getDirection` is already in Intl API, just not exposed, so it should not cause any cost to the vendors deploying it.

**Comment by zbraniecki** [comment]
I guess it's not the right idea to at the same time refactor a major piece of code and introduce new feature for debate in it :)

So I updated my `locale` branch to just introduce `IntlLocale` to `Intl.js` polyfill now - https://github.com/zbraniecki/Intl.js/tree/locale

@caridy - does it look like something you may be interested in?

**Comment by caridy** [comment]
_update: we need to cook the proposal for the next meeting in two weeks. will try to coordinate some time to work on this._

**Comment by zbraniecki** [comment]
We've talked about this with @caridy and he mentioned that Intl formatters may be interested in switching to something like `prioritizeLocales` for language negotiation.

This would mean that we don't have to expose `resolveLocale` at all, and just exposing `prioritizeLocales` will be enough for both - localization frameworks and for custom intl formatters to match the behavior of the platform ones.

The only thing coming to my mind that we may want to use `depth` parameter to limit the length of the resolved chain - if Intl formatter will only take the first one anyway, there's no reason to resolve the whole chain. At the same time l10n frameworks operate in more uncertainty (l20n for example falls back when a single l10n string cannot be resolved in a given language at runtime) so they need full "depth".

That would leave us with:

* isStructurallyValidLanguageTag
* canonicalizeLocaleList
* prioritizeLocales(availableLocales, requestedLocales, defaultLocale, depth)
* getDirection

(and later)
* `firstDayOfTheWeek` (or `weekendInfo` returning `firstDay`, `weekendStarts` and `weekendEnds`)

wrt. prioritizeLocales you can see an implementation here: https://github.com/l20n/l20n.js/blob/v1.0.x/lib/l20n/intl.js#L434

It uses `LookupAvailableLocales` which is a new function that acts exactly like `LookupSupportedLocales` but instead of returning a requested locale that has a match in available ones, it returns the available locale that has a match in requested ones.
So if the user requests 'sv-SE' and we have 'sv', `LookupSupportedLocales` will return 'sv-SE', while the new function will return 'sv'.  Which is what l10n and intl can later use to load the right resources.

**Comment by rxaviers** [comment]
I'm interested to know the exact algorithm and how this relates to https://github.com/rxaviers/ecma402-fix-lookup-matcher please.

**Comment by caridy** [comment]
@rxaviers that's precisely the reason why we will abstract the operation to produce the parent list, then we can update that to take into consideration the proper algo step.

**Comment by caridy** [comment]
[Abstract Locale Operations - Nov 2015](https://github.com/tc39/ecma402/files/41766/Abstract.Locale.Operations.pdf)

**Comment by caridy** [comment]
@rxaviers Ideally we can get https://github.com/rxaviers/ecma402-fix-lookup-matcher to fit into `getParentLocales()`, in other words, make the lookup matcher fix so we can expose it via `getParentLocales()`. Can you work with @zbraniecki to get that flushed out?

**Comment by rxaviers** [comment]
@caridy, note https://github.com/rxaviers/ecma402-fix-lookup-matcher is about finding the bundle, not the parent, which is the first entry of the chain. For example, given the locale `zh-TW`, you start at the `zh-Hant` bundle, and then inherits from `zh-Hant` parents, probably using  `getParentLocales()`.

**Comment by caridy** [comment]
yes @rxaviers, the fact that all those algos will require the inheritance path to make a decision about the data availability, that's the piece we want to expose, so they can do that logic in user-land as well.

**Comment by rxaviers** [comment]
It's in my TO-DO, but I got few spare time at the moment to go through it. Please, alert me in case I'm blocking you, but I will try to catch up time permitting. Thanks

**Comment by zbraniecki** [comment]
Here's the initial POC of the patch for Intl.js to implement the abstract locale operations discussed in the slides @caridy linked on Nov 23rd: https://github.com/andyearnshaw/Intl.js/compare/master...zbraniecki:locale

@caridy - does it look good? If so, I'll work on the spec patch.

**Comment by caridy** [comment]
@zbraniecki let's work on the spec, and then we can review the implementation. In principle, it looks good.

**Comment by zbraniecki** [comment]
Here's the first draft of the spec - https://github.com/tc39/ecma402/compare/master...zbraniecki:locale-api
I didn't write any algo for `resolveLocaleInfo` because I'm not sure how to approach accessing the necessary information from CLDR.

**Comment by rxaviers** [comment]
@caridy @zbraniecki sorry for the long delay on this reply and for my short worded comments above, but the end of the year has been intensive. First of all, great job with the Locale API, it's a good way of exposing these info and will definitely benefit custom libraries and polyfills.

How can I help you to get https://github.com/rxaviers/ecma402-fix-lookup-matcher into `getParentLocales()`? Note the first draft above gives wrong results [[details](https://gist.github.com/rxaviers/ee2d02719b4eb1dc0284)].

**Comment by zbraniecki** [comment]
That's for week data:
 * http://www.unicode.org/reports/tr35/tr35-dates.html#Week_Data 
 * http://www.unicode.org/reports/tr35/tr35.html#Key_Type_Definitions

**Comment by zbraniecki** [comment]
Heads up, this has been advanced to Stage 2 as of today. I'm going to work on updating the spec/polyfill now to @rxaviers and TC39 feedback and will submit it to reviewers in a couple weeks.

**Comment by stasm** [comment]
CLDR defines groups of supplemental data for each territory, like `timeData` or `weekData`.  An incomplete table can be consulted here:

http://www.unicode.org/cldr/charts/28/supplemental/territory_information.html

And the full data set here:

http://unicode.org/repos/cldr/trunk/common/supplemental/supplementalData.xml

Perhaps a good future-proof solution would be to base the method names on these groups, e.g.:

    Intl.Locale('fr-FR').getWeekData();

which would return:

    {
        minDays: {
            count: 4
        },
        firstDay: {
            day: 'mon'
        },
        weekendStart: {
            day: 'sat'
        },
        weekendEnd: {
            day: 'sun'
        }
    }

Or:
    
    Intl.Locale('fr-FR').getTimeData();

which would return:

    {
        hours: {
            preferred: 'H',
            allowed: 'H hB'
        }
    }
    
And for a different French-speaking territory:
    
    Intl.Locale('fr-CA').getTimeData();

which would return:

    {
        hours: {
            preferred: 'h',
            allowed: 'h hb H hB'
        }
    }

This is a bit verbose but it maps the CLDR data one-to-one which should make it much easier to extend it in the future.

**Comment by rxaviers** [comment]
FWIW, find CLDR in the JSON format here https://github.com/unicode-cldr/cldr-core/tree/master/supplemental.

**Comment by zbraniecki** [comment]
@caridy - do you need someone to write a patch for `Intl.js` to add `Intl.getCanonicalLocales` ?

**Comment by zbraniecki** [comment]
@caridy - also, why can't we get `Intl.getParentLocales` now? In the implementation for `Intl.js` we could start in a dummy form of cutting by `-` and we focus on updating the spec to use it everywhere. The addition of the exception then would be much easier in 4th ed.

**Comment by caridy** [comment]
@zbraniecki there is no time, and that one will require some extra work, where to get the data, etc, while the getCanonicalLocales is just a call to an abstract operation.

yes, a PR to Intl.js with it will be great.

**Comment by zbraniecki** [comment]
@caridy - https://github.com/andyearnshaw/Intl.js/pull/159

**Comment by zbraniecki** [comment]
Just a heads up - `Intl.getCanonicalLocales` just landed in Firefox ( https://bugzilla.mozilla.org/show_bug.cgi?id=1263040 ).

**Comment by zbraniecki** [comment]
I'm developing an internal API for Gecko which will be used by various date/time calendars and pickers.

At the moment the proposal for the API to retrieve the calendar data is:

```
let info = Intl.getCalendarInfo(locales);
```

which returns:
```
{
  "firstDayOfWeek": 7,
  "weekendStart": 6,
  "weekendEnd": 1,
  "locale": "de",
  "calendar": "gregory" 
}
```

The only nuance that I'm skipping for now is the concept of transition days where the weekend starts in the middle of the day and ICU can return UCAL_WEEKEND_ONSET and UCAL_WEEKEND_CEASE and then user can retrieve a corresponding time of the day when the transition occurs.

The reason I'm skipping it is because it doesn't seem to be used in CLDR data and can be added later. For example latest ICU uses `getWeekData` API which returns:
```
{
  weekendOnset: 6,
  weekendOnsetMills: ...,
  weekendCease: 1,
  weekendCeaseMills: ...
}
```

so we can also add `weekendStartTime`, and `weekendEndTime` if we decide later. I don't want to use `onset/cease` in place of `start/end` because I think it's more cryptic without any other value added.

As I said - this is an internal API, but maybe it'll help us design ECMA 402 API for that dataset.

**Comment by zbraniecki** [comment]
As I mentioned in #6 I landed `mozIntl.getCalendarInfo` for calendar information in Gecko.

Also, instead of `getDirection`, we're going to use `mozIntl.getLocaleInfo` which works very similarly to `getCalendarInfo`.

I'll be happy to revisit it and move away from that API toward any standard that ECMA402 will come up with.

**Comment by jacobrask** [comment]
Is there any further discussion on `Intl.getCalendarInfo` happening somewhere else? How did that API work out for you in Gecko?

**Comment by brettz9** [comment]
Some of the information in the IANA language subtag registry, a registry from which the following may be derived, may be [incomplete](https://stackoverflow.com/questions/47784379/why-are-some-prominent-encompassed-languages-missing-suppress-script), but I wonder whether the existing informaiton present could be used to supply `getLocaleInfo` with info on macrolanguages of encompassed languages?

This would, I imagine, be useful information for i18n as e.g., a user with language "in" (an encompassed language with parent macrolanguage "ms") could be directed to a "ms" locale if an "in" one could not be found.

**Comment by hugovdm** [comment]
I'm also looking for something like `resolveLocale` or `prioritizeLocales`: an API to call to help doing language matching (wanting to select a best locale based on the user's preferred list of locales and a developer-provided list of locales supported by an app, for example).

**Comment by opyh** [comment]
https://www.unicode.org/reports/tr35/#LanguageMatching proposes an algorithm to find a best fit for given sets of available and requested languages.

Could this be referenced in the spec - or maybe even integrated?

**Comment by sffc** [comment]
I see several discussions in this thread.  Let's move discussion of Intl.getCalendarInfo() over to #6, and use this thread to focus on the language matcher issues.

**Comment by johan** [comment]
> I'm developing an internal API for Gecko which will be used by various date/time calendars and pickers.
> 
> At the moment the proposal for the API to retrieve the calendar data is:
> 
> ```
> let info = Intl.getCalendarInfo(locales);
> ```
> 
> which returns:
> 
> ```
> {
>   "firstDayOfWeek": 7,
>   "weekendStart": 6,
>   "weekendEnd": 1,
>   "locale": "de",
>   "calendar": "gregory" 
> }
> ```
> 
> [...]
> 
> As I said - this is an internal API, but maybe it'll help us design ECMA 402 API for that dataset.

FWIW: if weekday references end up numbers instead of weekday name abbreviation names, as in CLDR (probably a good thing), it would be internally consistent to sync those numbers up with [ECMA-262's data type `Week Day`](https://tc39.es/ecma262/#sec-week-day) (the return value from [`Date.prototype.getDay`](https://tc39.es/ecma262/#sec-date.prototype.getday)), ranging from Sun=0 to Sat=6.

**Comment by ryzokuken** [comment]
`Intl.Locale` is Stage 4 and the follow-on Intl LocaleInfo is Stage 3. Can this be closed?

**Comment by sffc** [comment]
> `Intl.Locale` is Stage 4 and the follow-on Intl LocaleInfo is Stage 3. Can this be closed?

No.  As suggested in https://github.com/tc39/ecma402/issues/46#issuecomment-533336542, this issue is about language matching.  It would be fixed by the Intl.LocaleMatcher proposal.

https://github.com/tc39/proposal-intl-localematcher
